include/nbl/asset/utils/IShaderCompiler.h

```diff
@@ -157,32 +161,291 @@ class NBL_API2 IShaderCompiler : public core::IReferenceCounted
+							if (*passesIt != *otherPassesIt) return false;
+						}
+						return true;
+					}
+
+					inline std::vector<char> getHashable() const {
+						std::vector<char> hashable = preprocessorData.getHashable();
+						auto stageString = std::to_string(stage);
+						hashable.insert(hashable.end(), stageString.data()[0], stageString.data()[stageString.size()]);
+						auto versionString = std::to_string(static_cast<uint32_t>(targetSpirvVersion));
+						hashable.insert(hashable.end(), versionString.data()[0], versionString.data()[versionString.size()]);
+						auto debugString = std::to_string(static_cast<uint8_t>(debugInfoFlags.value));
+						hashable.insert(hashable.end(), debugString.data()[0], debugString.data()[debugString.size()]);
+						for (auto pass : optimizerPasses) {
+							auto passString = std::to_string(pass);
+							hashable.insert(hashable.end(), passString.data()[0], passString.data()[passString.size()]);
+						}
+						return hashable;
+					}
+
+					IShader::E_SHADER_STAGE stage;
+					E_SPIRV_VERSION targetSpirvVersion;
+					std::vector<ISPIRVOptimizer::E_OPTIMIZER_PASS> optimizerPasses;
+					core::bitflag<E_DEBUG_INFO_FLAGS> debugInfoFlags;
+					SPreprocessorData preprocessorData;
+				};
+
+				// The ordering is important here, the dependencies MUST be added to the array IN THE ORDER THE PREPROCESSOR INCLUDED THEM!
+				using dependency_container_t = std::vector<SPreprocessingDependency>;
+				inline SEntry(std::string_view _mainFileContents, dependency_container_t&& _dependencies, SCompilerData&& compilerData) : mainFileContents(std::move(std::string(_mainFileContents))), dependencies(std::move(_dependencies))
+				{
+				}
+
+				// Needed to get the vector deserialization automatically
+				SEntry() {}
+				inline SEntry(SEntry&&) = default;
+				inline SEntry& operator=(SEntry&&) = default;
+
+				// This next bit is a bit of a Frankenstein. We serialize shader creation parameters into a json, while the actual shader code goes in another file
+				struct CPUShaderCreationParams {
+					IShader::E_SHADER_STAGE stage;
+					IShader::E_CONTENT_TYPE contentType; //I think this one could be skipped since it's always going to be SPIR-V
+					std::string filepathHint;
+					uint64_t codeByteSize = 0;
+					uint64_t offset = 0; // Offset into the serialized .bin for the Cache where code starts
+				};
+
+				std::string mainFileContents;
+				SCompilerData compilerData;
+				// Keeping this one commented out. Could be useful for lazy-loading of dependencies if it takes up too much memory
+				// uint64_t entryID;
+
+				dependency_container_t dependencies;
+				mutable CPUShaderCreationParams shaderParams;
+				bool serialized = false;
+				mutable core::smart_refctd_ptr<asset::ICPUShader> value = nullptr;
+			};
+
+			inline void insert(SEntry&& entry)
+			{
+				m_container.insert(std::move(entry));
 			}
 
```

Comment:
I'd also add a 
```cpp
inline void merge(const CCache* other)
```

Reply:
and a `clone` that does a "deep copy" of the cache (with new/default `ICPUBuffer`s)

Reply:
do I add default move, deleted copy contructors?

Reply:
What do I do with constructors/assignment? Deleted copy, default move?

Reply:
things that inherit from `IReferenceCounted` always live on the heap, therefore they have no copy or move assignment nor constructors

Furthermore if their creation could fail, they need a non-public constructor and all the potentially failing parts of creation hoisted out to either a `static core::smart_refctd_ptr<this_t> create` factory method or some other facotry class method